### PR TITLE
Fix error handling in x509v3_cache_extensions and related functions

### PR DESCRIPTION
**Pull request by bernd-edlinger**
Basically we use EXFLAG_INVALID for all kinds of out of memory and
all kinds of parse errors in x509v3_cache_extensions.

[extended tests]

##### Checklist

- [ ] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

File: crypto/x509/x509_trs.c [hunk 1]
```diff
@@ -240,7 +240,8 @@ static int trust_1oid(X509_TRUST *trust, X509 *x, int flags)
 static int trust_compat(X509_TRUST *trust, X509 *x, int flags)
 {
     /* Call for side-effect of computing hash and caching extensions */
-    X509_check_purpose(x, -1, 0);
+    if (X509_check_purpose(x, -1, 0) != 1)
+        return X509_TRUST_UNTRUSTED;
     if ((flags & X509_TRUST_NO_SS_COMPAT) == 0 && x->ex_flags & EXFLAG_SS)
```
[the remark this answers is not in the document]
Comment (FdaSilvaYY) [reply crypto/x509/x509_trs.c]: While you are here ;).
Otherwise, it is a very good work 👍

[the remark this answers is not in the document]
Comment (bernd-edlinger) [reply crypto/x509/x509_trs.c]: It was on my to-do list for way too long ;)

File: crypto/x509/v3_purp.c [hunk 2]
```diff
@@ -386,8 +395,11 @@ static void x509v3_cache_extensions(X509 *x)
         BASIC_CONSTRAINTS_free(bs);
         x->ex_flags |= EXFLAG_BCONS;
     }
+    else if (i != -1) {
```
Comment (paulidale) [inline crypto/x509/v3_purp.c]: Style nit: on previous line.

File: apps/rehash.c [hunk 3]
```diff
@@ -274,11 +274,19 @@ static int do_file(const char *filename, const char *fullpath, enum Hash h)
     if (x->x509 != NULL) {
         type = TYPE_CERT;
         name = X509_get_subject_name(x->x509);
-        X509_digest(x->x509, evpmd, digest, NULL);
+        if (!X509_digest(x->x509, evpmd, digest, NULL)) {
+            BIO_printf(bio_err, "out of memory\n");
```
Comment (paulidale) [inline apps/rehash.c]: Is out of memory the only possible failure?

File: crypto/x509/v3_purp.c [hunk 4]
```diff
@@ -400,8 +412,11 @@ static void x509v3_cache_extensions(X509 *x)
         PROXY_CERT_INFO_EXTENSION_free(pci);
         x->ex_flags |= EXFLAG_PROXY;
     }
+    else if (i != -1) {
```
Comment (paulidale) [inline crypto/x509/v3_purp.c]: Style nit: on previous line.

File: apps/rehash.c (hunk 3 above)
Comment (bernd-edlinger) [reply apps/rehash.c]: I think, only on X509*_get_ext_d2i a syntax or OOM error can both be possible

File: crypto/x509/v3_purp.c (hunk 4 above)
Comment (bernd-edlinger) [reply crypto/x509/v3_purp.c]: Fixed. Thanks!

**Comment by bernd-edlinger** [comment]
Mind, lookin at #10756 as well?

**Comment by bernd-edlinger** [comment]
Merged to master  7e06a67
Thanks!

**Pull request closed without merging**